sql/sql_connect.cc

```diff
@@ -255,6 +261,125 @@ void time_out_user_resource_limits(THD *thd, USER_CONN *uc)
   DBUG_VOID_RETURN;
 }
 
+static uint64 get_connection_delay_time(uint32 count)
+{
+  /*
+   * Based on that invalid password errors will cause user can't connect server
+  ever and server close connection if it got message from client after
+  wait_timeout
+   * so ,
+   step = timeout/(max_password_errors - failed_attempts_before_delay)
+  delay = step *(failed_count  -failed_attempts_before_delay)
+*/
+  struct system_variables *p= &global_system_variables;
+  /* Time unit for delay step is seconds , convert to milliseconds*/
+  uint64 timeout= (p->net_wait_timeout < p->net_interactive_timeout
+                       ? p->net_wait_timeout
+                       : p->net_interactive_timeout) *
```

Comment:
I notice that read timeout value is 30 seconds , which means max delay is 30 seconds . Think from another point , client send login packet by network socket and server create a thread serving for connection. So in this  thread context only  when preceding instructions are executed ``` poll function will be called ``` to check if timeout.  Simplely first delay response and then check read timeout.  As a result , delay time and  timeout is independent.  Maybe I  dont't get full picture of vio framework. Would you give me some explanation? @vuvova

Reply:
Hmm, yes, you're right. The connection will not be automatically aborted because of the timeout. And indeed delay time and timeout are thus independent.

Still, `connect_timeout` (which is also `vio->read_timeout` and `vio->write_timeout` until the user is fully authenticated) is supposed to limit the duration of the login phase, so it kind of makes sense to artificially limit `delay` by `vio->read_timeout`. As a bonus it reduces the number of degrees of freedom, independent variables to configure, and makes two separate features parts of one larger feature